src/goto-programs/json_goto_trace.h

```diff
@@ -14,14 +14,140 @@ Date: November 2005
 #ifndef CPROVER_GOTO_PROGRAMS_JSON_GOTO_TRACE_H
 #define CPROVER_GOTO_PROGRAMS_JSON_GOTO_TRACE_H
 
+#include "goto_trace.h"
+
 #include <util/json.h>
+#include <util/json_stream.h>
+#include <util/json_expr.h>
+#include <util/arith_tools.h>
+#include <util/config.h>
+#include <util/invariant.h>
+#include <util/simplify_expr.h>
+#include <util/json_irep.h>
+#include <langapi/language_util.h>
 
-#include "goto_trace.h"
+/// This is structure is here to facilitate
+/// passing arguments to the conversion
+/// functions.
+typedef struct
+{
+  const jsont &location;
+  const goto_trace_stept& step;
+  const namespacet &ns;
+  const source_locationt &source_location;
+} conversion_dependenciest;
+
+/// Convert an ASSERT step.
+void convert_assert(json_objectt &json_failure,
+  const conversion_dependenciest &conversion_dependencies);
+
+/// Convert a DECL step.
+void convert_decl(json_objectt &json_assignment,
+ const conversion_dependenciest &conversion_dependencies,
+ const trace_optionst &trace_options);
 
+/// Convert an OUTPUT step.
+void convert_output(json_objectt &json_output,
+ const conversion_dependenciest &conversion_dependencies);
+
+/// Convert an INPUT step.
+void convert_input(json_objectt &json_input,
+ const conversion_dependenciest &conversion_dependencies);
+
+/// Convert a FUNCTION_RETURN step.
+void convert_return(json_objectt &json_call_return,
+  const conversion_dependenciest &conversion_dependencies);
+
+/// Handle all other kinds of steps that are not
+/// already handled by the other conversion functions.
+void convert_default(json_objectt &json_location_only,
+  const conversion_dependenciest &conversion_dependencies);
+
+/// Templated version of the conversion method.
+/// Works by dispatching to the more specialised
+/// conversion functions based on the type of the
+/// step that is being handled.
+/// \param ns: The namespace used for resolution of symbols.
+/// \param goto_trace: The goto_trace from which we are
+///   going to convert the steps from.
+/// \param dest_array: The JSON object that we are going
+///   to append the step information to.
+/// \param trace_options: A list of trace options
+template<typename json_arrayT>
 void convert(
-  const namespacet &,
-  const goto_tracet &,
-  jsont &,
-  trace_optionst trace_options = trace_optionst::default_options);
+  const namespacet &ns,
+  const goto_tracet &goto_trace,
+  json_arrayT &dest_array,
```

Comment:
Why does this need to be a template argument?

Reply:
To be able to use both the streaming and the object version.